### ISSUE TITLE
Run _createClass with object literals to allow property name mangling

### ISSUE THREAD

**Issue by ashi009** [babel/babel#10632]
## Feature Request

**Is your feature request related to a problem? Please describe.**

Assuming a simple pipeline like babel -> terser. When `transfrom-classes` is enabled, the generated code won't be able to be mangled by terser with `mangle.properties` option turned on.

In the transformed code `_createClass` is called with an array of descriptors to declare, and each descriptor of them has a `key` field with its original name. As it's a plain string, terser won't treat it as an object property. Once mangled, the caller side of the code will use the mangled property name, yet the declaring side will use the original name.

**Describe the solution you'd like**
A clear and concise description of what you want to happen. Add any considered drawbacks.

instead of calling _createClass with an array of descriptors, use an object literal instead:

```js
_createClass(Value, {
  method: {
    value: function method(name, error, value) {
      //...
    }
  }
}, {})
```

**Describe alternatives you've considered**
- Set up a different pipeline to mangle first then to babel. 
  It's less than ideal. To generate the minimized output, mangler needs to know the word frequency first, which needs babel to do its magic first.
- Teach mangler to mangle `key`s
  By adding some magic comments like `/*#__PURE__*/` to the `key`s, and teach manglers to treat them as property names. This is really a long shot, as there are so many tools around, and it will take ages to make the new magic string "de facto".

**Teachability, Documentation, Adoption, Migration Strategy**
Ideally, it should be a transparent change to the users, as this only changes the behavior of a plugin.

**Comment by nicolo-ribaudo** [comment]
Can you post a comparison of the mangled result for the current output and for the proposed output?

**Comment by ashi009** [comment]
Sure thing.


Sample class:

```js
export class A {
  methodA() {}
  methodB() {
   	console.log(this.methodA());
  }
}
```

The following snippets are the diffs of the current implementation and the patched one.

After Babel ([repl](https://babeljs.io/repl#?babili=false&browsers=last%201%20and_chr%20version%0A&build=&builtIns=usage&spec=false&loose=false&code_lz=KYDwDg9gTgLgBAYwDYEMDOa4EE4G8BQccAtsDABYQAmWAFAJR4C-hJZlVAQg3q3AJAIIAOzQQkwAHRIIAc1oUAlmkmkK1OvXoBuVixZA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=false&presets=es2015&prettier=false&targets=&version=7.6.4&externalPlugins=)):

```diff
@@ -6,15 +6,16 @@ function () {
     _classCallCheck(this, A);
   }

-  _createClass(A, [{
-    key: "methodA",
-    value: function methodA() {}
-  }, {
-    key: "methodB",
-    value: function methodB() {
-      console.log(this.methodA());
+  _createClass(A, {
+    methodA: {
+      value: function methodA() {}
+    },
+    methodB: {
+      value: function methodB() {
+        console.log(this.methodA());
+      }
     }
-  }]);
+  });

   return A;
 }();
```

After uglify

```diff
@@ -9,7 +9,7 @@ function _classCallCheck(e, n) {
 }
 
 function _defineProperties(e, n) {
-  for (var t = 0; t < n.length; t++) {
+  for (var t in n) if (Object.prototype.hasOwnProperty.call(n, t)) {
     var o = n[t];
     o.e = o.e || !1, o.n = !0, "value" in o && (o.t = !0), Object.defineProperty(e, o.o, o);
   }
@@ -20,22 +20,23 @@ function _createClass(e, n, t) {
 }
 
 Object.defineProperty(exports, "__esModule", {
-  r: !0
-}), exports.s = void 0;
+  i: !0
+}), exports.r = void 0;
 
 var A = function () {
   function e() {
     _classCallCheck(this, e);
   }
-  return _createClass(e, [{
-    o: "methodA",
-    r: function () { }
-  }, {
-    o: "methodB",
-    r: function () {
-      console.log(this.i());  // <-- `this.i` is undefined
+  return _createClass(e, {
+    s: {
+      i: function () { }
+    },
+    c: {
+      i: function () {
+        console.log(this.s()); // <-- `this.s` is defined above
+      }
     }
-  }]), e;
+  }), e;
 }();
 
-exports.s = A;
+exports.r = A;
```

After Closure Compiler:
```diff
@@ -2,12 +2,15 @@ Object.defineProperty(exports, "__esModule", { value: !0 });
 exports.a = void 0;
 function e(a, c) {
   if (c) {
-    for (var f = a.prototype, d = 0; d < c.length; d++) {
-      var b = c[d];
-      b.enumerable = b.enumerable || !1;
-      b.configurable = !0;
-      "value" in b && (b.writable = !0);
-      Object.defineProperty(f, b.key, b);
+    var f = a.prototype, d;
+    for (d in c) {
+      if (Object.prototype.hasOwnProperty.call(c, d)) {
+        var b = c[d];
+        b.enumerable = b.enumerable || !1;
+        b.configurable = !0;
+        "value" in b && (b.writable = !0);
+        Object.defineProperty(f, b.key, b);
+      }
     }
   }
 }
@@ -17,15 +20,17 @@ var g = function () {
       throw new TypeError("Cannot call a class as a function");
     }
   }
-  e(a, [{
-    key: "methodA",
-    value: function () { }
-  }, {
-    key: "methodB",
-    value: function () {
-      console.log(this.b()); // <- `this.b` is undefined, and closure compiler warned about it.
+  e(a, {
+    b: {
+      value: function () {
+      }
+    },
+    c: {
+      value: function () {
+        console.log(this.b()); // <- `this.b` is defined above.
+      }
     }
-  }]);
+  });
   return a;
 }();
 exports.a = g;
```

**Comment by jridgewell** [comment]
The issue with this is we'll have to use `Object.getOwnPropertySymbols` to get any symbol'd keys. Otherwise, this is fine.

**Comment by JLHwung** [comment]
We may replace `{ key: Identifier, value: function() }` by `[Identifier, function()]` without changing the helper logic too much.

Note that any helper interface change is a breaking change unless we go with a new helper such as `_defineProperty2`.

**Comment by ashi009** [comment]
> We may replace `{ key: Identifier, value: function() }` by `[Identifier, function()]` without changing the helper logic too much.
> 
> Note that any helper interface change is a breaking change unless we go with a new helper such as `_defineProperty2`.

This won't work either. A plain string cannot be naturally interpreted as a property. Also, it will break getter/setter definitions.

**Comment by ashi009** [comment]
> The issue with this is we'll have to use `Object.getOwnPropertySymbols` to get any symbol'd keys. Otherwise, this is fine.

Yep. Just out of curiosity, why not call `Object.defineProperties()` with all the descriptors, which already handles symbol'd keys.

**Comment by loganfsmyth** [comment]
Personally, I'd have a lot of reservations about this. Even if we change this one specific case, what about other transforms? For instance
```
var { first, ...rest } = {};
```
becomes
```
var _ref = {},
    first = _ref.first,
    rest = _objectWithoutProperties(_ref, ["first"]);
```
Do we also need to handle mangling of `first` too?

What happens if there are situations where we are unable to transform code in such a way that a mangler can tell what to do? For decorators for instance, it has at times been important that individual class elements evaluate in order. 

I worry that aiming to support this partially is still just as bad.

Babel hasn't generally aimed to make any guarantees about the output of plugins other than final runtime behavior, so changing that would be an increase in the surface area where we need to be concerned about breakage for any change we make.

**Comment by nicolo-ribaudo** [comment]
Also, it wouldn't work with this code:

```js
function getX() { return "x" }
function getX2() { return "x" }

class A {
  get [getX()]() {}
  set [getX2()](_) {}
}
```

Since using an object the `set` would overwrite the `get` instead of extending it.

**Comment by ashi009** [comment]
@loganfsmyth Sounds fair. 

Property name mangling has been a very challenging problem to tackle, given the dynamic nature of js.  However, anyone or any company that wants to do this will write their code against some strict style guidelines (like [Google's][1]), and even willing to avoid using certain language features.

My team writes code carefully to allow code optimizers to do its magic, and so far it only works if no downlevel was made to the code.  As babel is probably one of the best open-source transpiler available, we hope that transforms done to the AST are preserving those nuances.

For the rest transform, it can be easily done preserving the property nature:

```
var _ref = {},
    first = _ref.first,
    rest = _objectWithoutProperties(_ref, {first:true});
```

This is indeed a longshot, but I think it's totally doable, as long as we add tests to cover those cases. (Just think about how the `sideEffects` flag being adopted to make tree-shaking possible without AST analysis.)

@nicolo-ribaudo Yes, my demo code won't work for your case. But it's just a corner case to be covered in real code. For instance, by making `_createClass` take an array of prop descriptors: 

```js
_createClass(Ctor, [
  {
    method1: {},
    method2: {},
    [getX()]: {},
  }, {
    [getX2()]: {},
  }, {
    method3: {}
    [getX3()]: {},
  },
]);
```

[1]: https://google.github.io/styleguide/jsguide.html#features-objects-mixing-keys
